src/Analyser/ConstantResolver.php

```diff
@@ -302,8 +302,14 @@ public function resolvePredefinedConstant(string $resolvedConstantName): ?Type
 
 	public function resolveConstantType(string $constantName, Type $constantType): Type
 	{
-		if ($constantType->isConstantValue()->yes() && in_array($constantName, $this->dynamicConstantNames, true)) {
-			return $constantType->generalize(GeneralizePrecision::lessSpecific());
+		if ($constantType->isConstantValue()->yes()) {
+			if (array_key_exists($constantName, $this->dynamicConstantNames)) {
+				$phpdocTypes = $this->dynamicConstantNames[$constantName];
+				return $this->reflectionProviderProvider->getReflectionProvider()->getSignatureMapProvider()->getTypeFromString($phpdocTypes, null);
```

Comment:
You should use TypeStringResolver for this, not SignatureMapProvider.

Reply:
How would I get and instance of `TypeStringResolver` from here?